### ISSUE TITLE
Volume Management

### ISSUE THREAD

**Issue by smira** [siderolabs/talos#8367]
Closely related: #8016 

## Problem Statement

Talos Linux is not flexible in the way it manages volumes, it occupies the whole system disk, creating an EPHEMERAL partition covering 99% of the disk space. User disk management is fragile, requires extra steps to get it to work properly (mounting into the kubelet), doesn’t support wiping disks, etc. Talos does not detect properly various partition types which leads to wiping user data (e.g. Ceph bluestore).

There were following requests from the users/customers which can’t be addressed in the current design:

- running Talos network booted (i.e. with `STATE` / `EPHEMERAL` on `tmpfs`)
- running Talos installed e.g. to the SBC’s SD card, but mounting `/var` from an NVMe/SSD
- Azure VMs with directly attached NVMes will always install Talos to the network volume, so locally attached NVMe can’t be used for e.g. `containerd` state
- splitting `etcd` data directory to a separate disk
- performing wipe operations on the contents of the `/var`:
    - wiping system directories (e.g. `containerd` state)
    - wiping user data (e.g. `/var/mnt/foo`)
- disk encryption of the user disks (volumes)
- read-only user disks mounts (e.g. mounting my precious photo archive to the Talos machine, making sure that Talos never touches the contents)
- volume/disk management operations without reboot:
    - wiping user disks/volumes
    - wiping system areas
- creating additional user-managed partitions on the system disk:
    - `/data`
    - swap space
- container image cache storage
- some persistent across reboots logs storage (e.g. storing installation logs during staged upgrades)

The proposed design provides an option to solve the issues mentioned above.

## Groundwork

Before we move into volume management operations, there is some amount of work that needs to be done to improve the blockdevice management operations:

- Talos should quickly and easily detect various filesystem/partition types, including the most common ones, the ones that can be used in Kubernetes clusters. The minimum detection of the filesystem/partition prevents the disk from being considered empty and eligible for allocation
- The blockdevices/partitions should be presented as resources in a way that it allows to render them as a tree presenting a user with a view of available block devices, their types, partitions, filesystem types, available space, etc.
- Talos should detect and reliably show information which matches various standard Linux tools, e.g. `blkid`, to allow easier identification of storage objects.

## Installation Process

Talos installation should do the bare minimum to make sure that Talos can be booted from the disk, without touching the pieces which are not strictly required to boot Talos. This might include installing Talos without having machine configuration.

So the install should only touch following partitions/objects:

- `BOOT` / `EFI` partitions (boot assets, boot loader itself)
- `META` partition
- bootloader-specific stuff, e.g. writing to the `MBR`

Any management of the storage/volumes should be deferred to the Talos running on the `host` (i.e. creating `/var`, `/system/state`, etc.)

## Volumes

Let’s introduce a new concept of the `volumes`, which will solve the ideas mentioned above and allow us to take storage management to the next level.

There are two kinds of volumes:

- system volumes (i.e. required by Talos, and Talos provides default configuration for them if no other configuration is available)
- user volumes (configured and managed by users, optional)

Every volume has several most important features:

- lookup: Talos can find the volume by some selector, or say that the volume is not available
- provisioning (optional): if the volume is not available, Talos can provision the volume (e.g. create a partition), so that it can be looked up, and it becomes available
- parent: if the volume has a parent volume, it creates a dependency relationship between them
- mount path: might create another dependency on the volume which provides the mount path

Volumes support a basic set of operations:

- Mount/Unmount
- Wipe (requires unmounting first)
- Destroy (implies Wipe, but removes provisioned volumes)

Volume types:

- disk (e.g. use the whole disk)
- partition (allocate a partition on the disk)
- subdirectory (a sub-path on other Volume)

Volume formats:

- filesystem (or none)
- encryption (or none)
- …

Volume additional options:

- criticality (should be available for the pods to be started)
- mounted into the kubelet

### System Volumes

As of today, Talos implicitly has the following volume types:

| Name | Lookup | Provisioning | Format |
| --- | --- | --- | --- |
| STATE | a partition with the label STATE | create a partition on the system disk of size X MiB | xfs, optionally encrypted |
| EPHEMERAL | a partition with the label EPHEMERAL | create a partition on the system disk which occupies all remaining space | xfs, optionally encrypted |
| etcd data | - | subdirectory of EPHEMERAL, /var/lib/etcd | - |
| containerd data | - | subdirectory of EPHEMERAL, /var/lib/containerd | - |
| kubelet data | - | subdirectory of EPHEMERAL, /var/lib/kubelet | - |

### Volume Lifecycle

Talos services can express their dependency on the volumes. For example, `kubelet` service can only be started when `kubelet data` volume is available. Same way, if `kubelet data` volume is going to be unmounted, `kubelet` should be stopped first.

The boot process should naturally stop when the required volume is not available. E.g. maintenance mode of Talos implies that the boot can’t proceed as long as the volume configuration is not available.

### Volume Configuration

System volumes have implicit configuration, which is applied as long as `v1alpha1.Config` is applied to the machine. Some properties are configurable in `v1alpha1.Config`,  e.g. disk encryption. If an explicit volume configuration is provided, Talos uses that.

For example, if the user configures `EPHEMERAL` to be `tmpfs` of size 10 GiB, it will be created on each boot as instructed.

Users might provide configuration for user volumes (similar to the user disks feature today), which might be critical for the pods to be started, an otherwise e.g. extension services might provide a dependency on the additional volumes.

Some system volumes might be optional, i.e. configured by the users - for example, container image cache.

### Upgrades and Wiping

Talos Linux upgrades should not wipe anything by default, and wiping should be an additional operation which can be done without an upgrade, or can optionally be combined with an upgrade.

Update itself should only modify boot assets/boot loader, i.e. ensure that the new version of Talos Linux can be booted up from the disk device.

Wiping is volume-based, examples:

- I want to wipe `EPHEMERAL`, which implies wiping all volumes which have `EPHEMERAL` as a parent (e.g. subdirectory volume of `/var/lib/etcd`); all services which depend on `EPHEMERAL` or its children should be stopped, but reboot is not necessary, as the `EPHEMERAL` will be re-provisioned after the wipe
- I want to wipe `etcd data`, which in the default configuration implies leaving `etcd`, stopping `etcd` services, performing `rm -rf /var/lib/etcd`, and re-starting `etcd` join process

## Notes

As pointed out by @utkuozdemir, `EPHEMERAL` might be a bad name given that the partition is not supposed to be forced wiped by default.
```[tasklist]
### Tasks
- [ ] https://github.com/siderolabs/go-blockdevice/issues/78
- [ ] https://github.com/siderolabs/talos/issues/9261
- [ ] https://github.com/siderolabs/talos/issues/9471
- [ ] remove v1 go-blockdevice
- [ ] https://github.com/siderolabs/talos/issues/9602
- [ ] implement volume wipe
- [ ] https://github.com/siderolabs/talos/issues/9731
- [ ] https://github.com/siderolabs/talos/issues/9724
- [x] implement directory volumes
- [x] user volumes
- [ ] https://github.com/siderolabs/talos/issues/9530
```

**Comment by smira** [comment]
This feature is going to be shifted to Talos 1.8.0 (only first bits might appear in Talos 1.7.0). 

Talos 1.8.0 will be released as soon as this feature is ready.

**Comment by runningman84** [comment]
Some software like longhorn might not respect limits and fill the whole disk. It would be great if a misbehaving pod cannot destroy etcd or other core parts of talos by just claiming all the available disk space.

**Comment by andrewrynhard** [comment]
> Some software like longhorn might not respect limits and fill the whole disk. It would be great if a misbehaving pod cannot destroy etcd or other core parts of talos by just claiming all the available disk space.

This is good to know. I always liked the separation of Talos and it using a dedicated disk to prevent unknowns/complications like this. Any ideas on how we could impose those limitations?

**Comment by runningman84** [comment]
From my point of view something like lvm and partitions for each part would help. I used a similar setup in k3s and had never issues like this.

**Comment by bplein** [comment]
Allow the choice of any block device.  

A partition is also a block device.  People could partition a single SSD with sufficient space for Talos and then an additional partition for general use.  Filling up the general use partition isn’t going to affect the Talos partition(s)

**Comment by PeterFalken** [comment]
> Allow the choice of any block device.
> 
> A partition is also a block device. People could partition a single SSD with sufficient space for Talos and then an additional partition for general use. Filling up the general use partition isn’t going to affect the Talos partition(s)

Similar to how the newer ESXi installer does when using the `systemMediaSize` option, this allows the installer to make the system & OS partitions at the beginning of the disk.While leaving free space at the end of the disk.

**Comment by runningman84** [comment]
I think at minimal we would need two partitions or lvm volumes:
Talos (etcd and other stuff)
General purpose (container data and so on)

It would be great if we could have an option to say okay we also need 100gb longhorn space and 50gb local path space. That are just some examples we would just need a volume size and mount path. All remaining space could be assigned to the general purpose partition. Here the default setting should be to use all space.

With something like lvm we could also allow to fix the general volume to a specific size and leave the remaining space unused. The would allow for expansion of other volumes or ensure that all nodes are the same even if one has a bigger disk.

**Comment by cortopy** [comment]
> This feature is going to be shifted to Talos 1.8.0 (only first bits might appear in Talos 1.7.0).
> 
> Talos 1.8.0 will be released as soon as this feature is ready.

Thank you for clarifying @smira! If I set up a cluster with 1.7 today, will there be a migration path in 1.8 to have talos managing disks as proposed in this issue?

**Comment by smira** [comment]
> Thank you for clarifying @smira! If I set up a cluster with 1.7 today, will there be a migration path in 1.8 to have talos managing disks as proposed in this issue?

Talos is always backwards compatible, so upgrade to 1.8 will always work. You would be able to start using volume management features, but some of them (e.g. shrinking `/var`) might require a wipe of some volumes.

**Comment by chr0n1x** [comment]
this is incredibly exciting, happy to give it a whirl once you get an RC/beta or something  @smira . thank you!

**Comment by Ornias1993** [comment]
@smira TopoLVM requires a `pvcreate` and `vgcreate` to be able to allocate remaining free disk space.

What I get from this issue, is that we would be able to at least allocate the system disk with free-space remaining, which is already 99% of the way there!

Does it also allow us to use lvm using `pvcreate` and `vgcreate` to consume the rest of the disk space?
Note: not a lvm specialist at all.

**Comment by smira** [comment]
## Progress Update (1.8.0-alpha.2)

As soon as #8901 is merged, we will release the last alpha version.

At this point, the internals for volume management got completely implemented from scratch, while there are new configuration options available to use new features.

Before 1.8.0-beta.0, we plan to add minimal configuration for `EPHEMERAL` - creating on a different disk, controlling maximum/minimum size, and growth.

All other changes got shifted towards 1.9.0.

**Comment by schneid-l** [comment]
Hi @smira, huge congratulations for your incredible work on this PR 🎉 

Having a quick look at the content of the PR + the documentation added to it, I couldn't find an answer to the following question:

With Talos v1.8 will we be able to do a full-ram PXE boot without having to install the OS on a disk? Or is this one of the features shifting to v1.9?

Thanks a lot 🙏

**Comment by smira** [comment]
> With Talos v1.8 will we be able to do a full-ram PXE boot without having to install the OS on a disk? Or is this one of the features shifting to v1.9?

not in 1.8 (probably 1.9), two pieces are missing:

* ability to skip install itself (will come separate via Install API)
* `tmpfs` volumes for required EPHEMERAL/STATE (will come as part of this workflow)

**Comment by qjoly** [comment]
Well done on the work done on 1.8.0-alpha.2, can't wait to see the results in practice 👏 🎉 (Your newly added `get discoveredvolume` will help us a lot) 

If I understand correctly, the changes made mainly concern the new concept of volumes (which will be used for the new features in 1.9). This means that Machine Configurations are not yet impacted, am I right? 

Thank you in advance 🙏

**Comment by smira** [comment]
> If I understand correctly, the changes made mainly concern the new concept of volumes (which will be used for the new features in 1.9). This means that Machine Configurations are not yet impacted, am I right?

yes, for 1.8.0-alpha.2, there are no machine configuration changes, and everything "should just work as before".

Before 1.8.0-beta.0, there will be new machine configuration document to configure some aspects of `EPHEMERAL` volume (this is the most requested feature).

Everything else for 1.9.

**Comment by pitabwire** [comment]
@smira with the release of 1.8 can we have a user defined /data directory on the primary system disk or this will be done in the 1.9 release? I was looking at the docs and can't see how to create additional user-managed partitions on the system disk

**Comment by smira** [comment]
> @smira with the release of 1.8 can we have a user defined /data directory on the primary system disk or this will be done in the 1.9 release? I was looking at the docs and can't see how to create additional user-managed partitions on the system disk

With the release of 1.8 (or technically right now with v1.8.0-beta.0), you can shrink `EPHEMERAL` to be fixed size, located on a system disk or not.

There is no support for adding extra volumes besides legacy `machine.disks`.

1. You can add `/data` yourself (and Talos won't touch it)
2. You might not have immediate need `/data` anymore, as Talos never wipes on upgrades.

More user volume support is planned for 1.9.

**Comment by smira** [comment]
Current state of things for volume/disk management: https://www.talos.dev/v1.8/talos-guides/configuration/disk-management/

**Comment by isometry** [comment]
Given the new implementation and support for configuring the EPHEMERAL volume, is it now possible to split a disk between `machine.disks` and VolumeConfig?
If I configure `machine.disks[0].partitions[0].size` somewhat smaller than the associated NVMe drive, and then configure a VolumeConfig with diskSelector targeting the same drive, will it automagically add and use a second partition for the EPHEMERAL volume? (Use case: getting the most out of internal NVMe in a homelab Turing Pi 2 cluster with Longhorn…)

**Comment by smira** [comment]
> Given the new implementation and support for configuring the EPHEMERAL volume, is it now possible to split a disk between `machine.disks` and VolumeConfig? If I configure `machine.disks[0].partitions[0].size` somewhat smaller than the associated NVMe drive, and then configure a VolumeConfig with diskSelector targeting the same drive, will it automagically add and use a second partition for the EPHEMERAL volume? (Use case: getting the most out of internal NVMe in a homelab Turing Pi 2 cluster with Longhorn…)

yes, it should work I believe

**Comment by nadenf** [comment]
Was asked to add some notes about software RAID for on-premise i.e. metal. Not looking for a response just wanted to add another viewpoint.

1) I appreciate that the goal of Kubernetes is for disposable hardware. But that’s not the reality these days where GPUs are expensive and hard to come by. So for us we would rather not run workloads as opposed to spend another 20k just to have a backup GPU server sitting there idle in case of failure. So any improvement in per-node reliability is a big deal for us.

2) Hardware RAID is not an option for Intel hardware as VROC/DST only works with Intel SSDs. And using third party cards is difficult because the format is often proprietary. So if you can’t get easily get a replacement card your data is lost.

3) Using OpenEBS, Ceph etc is an issue because we aren’t looking to replicate data across multiple machines. This cripples performance where our data sizes are large and we are sensitive to latency. We just want the data on our node to be more available. Also it’s not exactly a simple piece of software to operate.

**Comment by PrivatePuffin** [comment]
@nadenf I dont get what you aim for here with this response.
It adds nothing to the issue, which is just about adding volume management.


1. Completely irrelevant to this issue about volume management.
2. Ok, so what does this has to do with hardware raid? Nothing at all.
3.a. OpenEBS LocalPV is not a replicated form of storage.
3.b. OpenEBS ALSO offers a ZFS backend
3.c. this issue has absolutely NOTHING to do with availability in itself

**Comment by nadenf** [comment]
>@nadenf I dont get what you aim for here with this response.
 It adds nothing to the issue, which is just about adding volume management.

I was asked to add it here. By Steve Francis, the CEO of Sidera Labs.

And perhaps the thinking is that since a RAID is a form of volume that the design whilst not supporting software RAID could at least be done in a way that does not prevent it being added in the future.

**Comment by PrivatePuffin** [comment]
> I was asked to add it here. By Steve Francis, the CEO of Sidera Labs.

I bet you also emailed him some off-topic crap out-of-the-blue like you did here.

> And perhaps the thinking is that since a RAID is a form of volume that the design whilst not supporting software RAID could at least be done in a way that does not prevent it being added in the future.

In that case, freaking say that.
Instead of rambling on about all sorts of things that that have nothing to do with the issue.

**Comment by 0dragosh** [comment]
Just wanted to pop in and say the entire community loves Talos and just ignore the one butthead.

Keep up the good work guys!

**Comment by absolutejam** [comment]
> Given the new implementation and support for configuring the EPHEMERAL volume, is it now possible to split a disk between `machine.disks` and VolumeConfig? If I configure `machine.disks[0].partitions[0].size` somewhat smaller than the associated NVMe drive, and then configure a VolumeConfig with diskSelector targeting the same drive, will it automagically add and use a second partition for the EPHEMERAL volume? (Use case: getting the most out of internal NVMe in a homelab Turing Pi 2 cluster with Longhorn…)

Did you have any joy with this approach @isometry? Would love to see an example if so!

**Comment by maxpain** [comment]
It would be helpful to create custom partitions.
In my case, my blades have two HDDs for Ceph and only one M.2 NVMe for Talos Linux. I want to use this SSD not only for Talos but also for Ceph metadata, but it requires a dedicated block device, so it's not possible to use `hostath`.

**Comment by smira** [comment]
It's designed to be working this way. Keep in mind that talosctl reset without specific labels will blow up the whole disk still, but we are working on a solution for that as well.

**Comment by maxpain** [comment]
> It would be helpful to create custom partitions. In my case, my blades have two HDDs for Ceph and only one M.2 NVMe for Talos Linux. I want to use this SSD not only for Talos but also for Ceph metadata, but it requires a dedicated block device (it's impossible to use `hostPath`).

Unfortunately, Ceph doesn't support the use of partitions for OSD:
```
rookcmd: failed to configure devices: failed to initialize osd: Partition device nvme0n1p7 can not be specified as metadataDevice in the global OSD configuration or in the node level OSD configuration
```

It would be cool to have a way to manage NVMe namespaces in the machine configuration.

**Comment by davralin** [comment]
> > It would be helpful to create custom partitions. In my case, my blades have two HDDs for Ceph and only one M.2 NVMe for Talos Linux. I want to use this SSD not only for Talos but also for Ceph metadata, but it requires a dedicated block device (it's impossible to use `hostPath`).
> 
> 
> 
> Unfortunately, Ceph doesn't support the use of partitions for OSD:
> 
> ```
> 
> rookcmd: failed to configure devices: failed to initialize osd: Partition device nvme0n1p7 can not be specified as metadataDevice in the global OSD configuration or in the node level OSD configuration
> 
> ```
> 
> 
> 
> It would be cool to have a way to manage NVMe namespaces in the machine configuration.

I have previously used partitions for rook, so I know this _used_ to work, but that was with kubeadm-clusters, so things might have changed since I switched to talos 2 years ago.

That being said, [this](https://rook.io/docs/rook/latest-release/CRDs/Cluster/ceph-cluster-crd/?h=lvm#storage-selection-settings) mentions partitions several times as something one can use.

**Comment by seang96** [comment]
My cluster runs nixos right now waiting for this feature on Talos. I specify by disk uuid for ceph though not the partition name and it works fine.

**Comment by maxpain** [comment]
> That being said, [this](https://rook.io/docs/rook/latest-release/CRDs/Cluster/ceph-cluster-crd/?h=partition#storage-selection-settings) mentions partitions several times as something one can use.

Hmm.

<img width="812" alt="image" src="https://github.com/user-attachments/assets/502f21f2-d015-42cc-a740-c05ff7afb149" />

**Comment by halittiryaki** [comment]
after shrinking the ephemeral disk, how can i mount the remaining disk space as seperate partition to eg: `/var/mnt/persistent`

```yaml
apiVersion: v1alpha1
kind: VolumeConfig
name: EPHEMERAL
provisioning:
  diskSelector:
    match: system_disk
  minSize: 200GB
  maxSize: 200GB
  grow: false
```

the `disks` section in machineconfig only is able to select devices,.
so the only options currently would be to use staticpods/daemonsets that calls `mount ..` for a manually created partition on the remaining space?

**Comment by halittiryaki** [comment]
> Any workarounds for now? I set a fixed size for `EPHEMERAL` volume using `VolumeConfig`:
> 
> ```yaml
> apiVersion: v1alpha1
> kind: VolumeConfig
> name: EPHEMERAL
> provisioning:
>   maxSize: 30GiB
>   grow: false
> ```
> 
> And then manually created a custom volume for Ceph metadata:
> 
> ![telegram-cloud-photo-size-2-5359480000241330475-x](https://private-user-images.githubusercontent.com/5875582/391870493-f969851b-d568-4979-ba00-03542f86596b.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyOTM0MTYsIm5iZiI6MTczNDI5MzExNiwicGF0aCI6Ii81ODc1NTgyLzM5MTg3MDQ5My1mOTY5ODUxYi1kNTY4LTQ5NzktYmEwMC0wMzU0MmY4NjU5NmIuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTVUMjAwNTE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWI4NWQ5OTkxYzkxYTZjYzMzMGQxNjExZDljZjVmZjlkZjIyNzgwZjYwZjM2ODFhYTVhNWQxMTA1MzFiYjI4YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.1cDFqHKK6m8_EROlc5-YJH6fsS0etH1TtIKHt7Cufmc)
> 
> How bad is this as a temporary solution? I even tried `talosctl upgrade`, and it works.

did you manage to manually create and bind the volume with talos?

**Comment by maxpain** [comment]
> did you manage to manually create and bind the volume with talos?

No, each node in my cluster has two HDDs for Ceph data and one NVMe SSD for Talos and Ceph metadata.
I decided to use NVMe namespaces instead. It allows me to have two or more "virtual" block devices instead of partitions, so I no longer need `VolumeConfig`.

**Comment by isometry** [comment]
> Did you have any joy with this approach @isometry? Would love to see an example if so!

Yes, I've (only just) finished testing this successfully.

I applied the following configuration snippets to a freshly reset Turing RK1 with [Talos v1.9.0](https://factory.talos.dev/image/85f683902139269fbc5a7f64ea94a694d31e0b3d94347a225223fcbd042083ae/v1.9.0/metal-arm64.raw.xz) installed to the MMC, and with a single, unpartitioned 1TB NVMe drive attached at `/dev/nvme0n1`:

```yaml
machine:
  # ...elided...
  disks:
    - device: /dev/nvme0n1
      partitions:
        - mountpoint: /var/lib/longhorn
          size: 800GB
  # ...elided...
  kubelet:
    extraMounts:
      - destination: /var/lib/longhorn # Destination is the absolute path where the mount will be placed in the container.
        type: bind # Type specifies the mount kind.
        source: /var/lib/longhorn # Source specifies the source path of the mount.
        # Options are fstab style mount options.
        options:
          - rbind
          - rshared
          - rw
  # ...elided...
---
apiVersion: v1alpha1
kind: VolumeConfig
name: EPHEMERAL
provisioning:
  diskSelector:
    match: disk.transport == "nvme"
  maxSize: 128GB
  grow: false
```

This results in the following volume configuration:

```console
$ talosctl get volumestatus
NODE         NAMESPACE   TYPE           ID               VERSION   PHASE   LOCATION         SIZE
turingpi-1   runtime     VolumeStatus   /dev/nvme0n1-1   2         ready   /dev/nvme0n1p2   800 GB
turingpi-1   runtime     VolumeStatus   EPHEMERAL        1         ready   /dev/nvme0n1p1   128 GB
turingpi-1   runtime     VolumeStatus   META             2         ready   /dev/mmcblk0p4   1.0 MB
turingpi-1   runtime     VolumeStatus   STATE            3         ready   /dev/mmcblk0p5   105 MB

$ talosctl mounts | egrep 'NODE|nvme'
NODE         FILESYSTEM       SIZE(GB)   USED(GB)   AVAILABLE(GB)   PERCENT USED   MOUNTED ON
turingpi-1   /dev/nvme0n1p1   127.93     5.44       122.50          4.25%          /var
turingpi-1   /dev/nvme0n1p2   871.78     16.76      855.01          1.92%          /var/lib/longhorn
```

I was a little surprised that the `EPHEMERAL` volume is partitioned before the machine disk, but it's fine. I also tested with multiple `machine.disks[].partitions[]`, and this works much the same.

Longhorn is happily running atop the `/var/lib/longhorn` mount.

**Comment by isometry** [comment]
Warning: following my previously described configuration, local storage is broken by an upgrade (to v1.9.1) :-/

```console
$ talosctl -n 10.0.88.11 get volumestatus
NODE         NAMESPACE   TYPE           ID               VERSION   PHASE   LOCATION         SIZE
turingpi-1   runtime     VolumeStatus   /dev/nvme0n1-1   1         ready   /dev/nvme0n1p1   128 GB
turingpi-1   runtime     VolumeStatus   EPHEMERAL        1         ready   /dev/nvme0n1p1   128 GB
turingpi-1   runtime     VolumeStatus   META             2         ready   /dev/mmcblk0p4   1.0 MB
turingpi-1   runtime     VolumeStatus   STATE            2         ready   /dev/mmcblk0p5   105 MB
```

I suspect that it is necessary to pre-partition the NVMe for all `machine.disks[].partitions[]` *before* applying any `VolumeConfig` patches to avoid this... Will try this now.

**Comment by smira** [comment]
User disks `.machine.disks` are not compatible with random placement of EPHEMERAL partition. They will eventually be deprecated and replace with a better mechanism.

**Comment by Blarc** [comment]
I might be a bit too early, and this might not be implemented yet, but I tried following to what @isometry wrote:

> > Did you have any joy with this approach @isometry? Would love to see an example if so!
> 
> Yes, I've (only just) finished testing this successfully.
> 
> I applied the following configuration snippets to a freshly reset Turing RK1 with [Talos v1.9.0](https://factory.talos.dev/image/85f683902139269fbc5a7f64ea94a694d31e0b3d94347a225223fcbd042083ae/v1.9.0/metal-arm64.raw.xz) installed to the MMC, and with a single, unpartitioned 1TB NVMe drive attached at `/dev/nvme0n1`:
> 
> ```yaml
> machine:
>   # ...elided...
>   disks:
>     - device: /dev/nvme0n1
>       partitions:
>         - mountpoint: /var/lib/longhorn
>           size: 800GB
>   # ...elided...
>   kubelet:
>     extraMounts:
>       - destination: /var/lib/longhorn # Destination is the absolute path where the mount will be placed in the container.
>         type: bind # Type specifies the mount kind.
>         source: /var/lib/longhorn # Source specifies the source path of the mount.
>         # Options are fstab style mount options.
>         options:
>           - rbind
>           - rshared
>           - rw
>   # ...elided...
> ---
> apiVersion: v1alpha1
> kind: VolumeConfig
> name: EPHEMERAL
> provisioning:
>   diskSelector:
>     match: disk.transport == "nvme"
>   maxSize: 128GB
>   grow: false
> ```
> 
> This results in the following volume configuration:
> 
> ```
> $ talosctl get volumestatus
> NODE         NAMESPACE   TYPE           ID               VERSION   PHASE   LOCATION         SIZE
> turingpi-1   runtime     VolumeStatus   /dev/nvme0n1-1   2         ready   /dev/nvme0n1p2   800 GB
> turingpi-1   runtime     VolumeStatus   EPHEMERAL        1         ready   /dev/nvme0n1p1   128 GB
> turingpi-1   runtime     VolumeStatus   META             2         ready   /dev/mmcblk0p4   1.0 MB
> turingpi-1   runtime     VolumeStatus   STATE            3         ready   /dev/mmcblk0p5   105 MB
> 
> $ talosctl mounts | egrep 'NODE|nvme'
> NODE         FILESYSTEM       SIZE(GB)   USED(GB)   AVAILABLE(GB)   PERCENT USED   MOUNTED ON
> turingpi-1   /dev/nvme0n1p1   127.93     5.44       122.50          4.25%          /var
> turingpi-1   /dev/nvme0n1p2   871.78     16.76      855.01          1.92%          /var/lib/longhorn
> ```
> 
> I was a little surprised that the `EPHEMERAL` volume is partitioned before the machine disk, but it's fine. I also tested with multiple `machine.disks[].partitions[]`, and this works much the same.
> 
> Longhorn is happily running atop the `/var/lib/longhorn` mount.

But I always get:
```
192.168.1.16: user: warning: [2024-12-28T18:25:29.291534284Z]: [talos] volume status {"component": "controller-runtime", "controller": "block.VolumeManagerController", "volume": "/dev/sda-1", "phase": "failed -> failed", "error": "filesystem type mismatch: vfat != xfs", "location": "/dev/sda1", "parentLocation": "/dev/sda"}
```
The EPHEMERAL is correctly sized to 100Gb, but the following machine mount does not work:
```
machine:
  disks:
      - device: /dev/sda
        partitions:
          - mountpoint: /var/local
            size: 400GB
```

**Comment by isometry** [comment]
> I might be a bit too early, and this might not be implemented yet, or I completely miss understood how this works, but I tried following to what @isometry wrote:
> ...
> The EPHEMERAL is correctly sized to 100Gb, but the following machine mount does not work:
> ...

@Blarc  check my follow-up warning. Ensure that your target disk has a GPT label and a single 400GB partition (which will be used for your `machine.disks` configuration) *before* you apply the initial Talos configuration (I personally achieved this by using a debug container on an already initialised node, then resetting just system disks). During initialisation, the VolumeConfig doc should create a second EPHEMERAL partition according to your spec. Note that you'll need to fully wipe/reset the system first to ensure the existing EPHEMERAL partition is removed; you do not want it as the first partition on your disk as that *will* collide with your `machine.disks` config when you upgrade.

**Comment by Blarc** [comment]
> > I might be a bit too early, and this might not be implemented yet, or I completely miss understood how this works, but I tried following to what @isometry wrote:
> > ...
> > The EPHEMERAL is correctly sized to 100Gb, but the following machine mount does not work:
> > ...
> 
> @Blarc check my follow-up warning. Ensure that your target disk has a GPT label and a single 400GB partition (which will be used for your `machine.disks` configuration) _before_ you apply the initial Talos configuration (I personally achieved this by using a debug container on an already initialised node, then resetting just system disks). During initialisation, the VolumeConfig doc should create a second EPHEMERAL partition according to your spec. Note that you'll need to fully wipe/reset the system first to ensure the existing EPHEMERAL partition is removed; you do not want it as the first partition on your disk as that _will_ collide with your `machine.disks` config when you upgrade.

Is there any chance you could provide an example how to ensure disk has GPT label?

**Comment by isometry** [comment]
> Is there any chance you could provide an example how to ensure disk has GPT label?

I don't think GPT or not matters. Just make sure that your disk contains a single 400GB partition before you apply your configuration (most easily done by resetting the host, applying *just* the `machine.disks` configuration, then *only* resetting the system-disk, and applying *with* the `VolumeConig` doc next time).

**Comment by rserbitar** [comment]
So is there any documentation (except the template in the .yaml config file) on how to use the machine.disks and how to interact with default EPHEMERAL? Or is this currently in testing state anyways? Setting Talos up in a way @isometry recommends (meaning requiring a certain order of commands to initialize) is kind of exactly what Talos wants to avoid, right?

**Comment by isometry** [comment]
@rserbitar "quite". From what I can tell, the machine disks mechanism is strictly *not* currently appropriate for trying to create or manage non-system partitions on the system disk. The longhorn/ephemeral on NVMe configuration I found to work is sub-optimal, as you say, in that it requires multiple configuration passes, and only works at all when you have (at least) a second disk for machine disks (I'm using the Turing RK1 embedded MMC for Talos system/meta/state, on the understanding that these should be almost read-only during normal operation).

**Comment by Syntax3rror404** [comment]
This also not covers second disks for data only, for example first nvme for talos and second for longhorn.

**Comment by isometry** [comment]
> This also not covers second disks for data only, for example first nvme for talos and second for longhorn. 

Use [diskSelector](https://www.talos.dev/v1.9/talos-guides/configuration/disk-management/#disk-selector) to differentiate between drives.

**Comment by Syntax3rror404** [comment]
This doesent work for machine.disks only for the install part.

**Comment by bra-fsn** [comment]
I'm completely new to Talos and considered it for a project where I would like to use ZFS and netboot.
After looking around I came to the conclusion that it's a no go for me, although I pretty much liked the other parts. :(
I may have misunderstood things, but wouldn't the following scheme solve most of the problems?

1. give the users two hooks: mount, init (coming from the yamls, so it can be flexible, like handle the differences between different types of underlying hardware).
1.1. mount could be used to mount arbitrary file systems to either the desired (baked-in) or flexible (can be changed in the yaml, so Talos could use different dirs) locations.
1.2. if mount fails (this is the first time the machine starts up), init could be run, which could create the necessary file systems (and on success, retries **1.1**)
2. provide "install-less" (auto-discovery) mode (for netbooting/ISO-booting etc). This should check for an already established talos instance, for example by looking for a marker file (like the current `node-identity.yaml`?). This should run after **1.1**/**1.2**.
2.1. if it sees an empty storage area (no marker file), it could set up everything and continue as normal
2.2. if it sees an already set up storage, it could just use it
2.3. on version change, it could do schema transitions as well if needed

I think thous would be flexible enough to cover most of the needs. I don't know enough about Talos to understand how hooks would look like (shell scripts for example) and where they would run.
Looking around inside the squashfs it seems it would be easier to run (some of) these hooks in privileged containers, but I'm not sure if it would cover everything (for example mounting).

**Comment by Ulexus** [comment]
Planning-wise, now that `machine.disks` is deprecated, do we need to worry that it might be removed before we can specify an alternative volume for `etcd` data?

As far as I can tell, we have no way of doing that any more, with the current version (using only non-deprecated features).

**Comment by smira** [comment]
> Planning-wise, now that `machine.disks` is deprecated, do we need to worry that it might be removed before we can specify an alternative volume for `etcd` data?

We do not plan to remove `.machine.disks` completely, and we do plan to offer a "proper" way to specify a dedicated volume for containerd and etcd.

**Comment by smira** [comment]
> Any thoughts on this? It would be nice if TopoLVM is supported as well.

We don't support LVM management done by Talos itself, but LVM volumes created work just fine, e.g. Ceph with encryption would use them without any issues on Talos.

**Comment by TsengSR** [comment]
Anything new on this topic? 

I'm also kinda having an issues with that. I'd like to run longhorn on my homelab, but I don't feel very comfortable having longhorn on the same EPIPHERMAL partition as talos, as one wrong command just wipes it on upgrade. 

Using multiple disks is out of question, the system is a Turing PI with 4 RK1 modules (Rockchip RK3588 with 32 GB RAM) and only one M.2 Slot per module on which talos is installed. The only other option would be to install Talos on the eMMC which I also don't feel very comfortable with as its only 32 GB and I don't wanna wear it out with images pulled for

**Comment by smira** [comment]
> Anything new on this topic?

Not sure what do you mean, parts of the whole story shipped in Talos 1.10, more in 1.11.
 
> I'm also kinda having an issues with that. I'd like to run longhorn on my homelab, but I don't feel very comfortable having longhorn on the same EPIPHERMAL partition as talos, as one wrong command just wipes it on upgrade.

Talos DOES NOT WIPE on upgrades anymore (since [Talos 1.8](https://www.talos.dev/v1.8/introduction/what-is-new/#upgrades) ). (Sorry for capital letters, but I want to stress this fact). So `EPHEMERAL` is safe, but you can also allocate dedicated partition/reduce EPHEMERAL size, etc.

Please see [documentation](https://www.talos.dev/v1.10/talos-guides/configuration/disk-management/) for further details.

**Comment by isometry** [comment]
> Anything new on this topic?

My setup is also built on Turing RK1 with an NVMe per node. YMMV, but I've been very happy with Talos imaged to the eMMC, then EPHEMERAL and Longhorn on the NVMe with the following:

```yaml
---
apiVersion: v1alpha1
kind: VolumeConfig
name: EPHEMERAL
provisioning:
  diskSelector:
    match: disk.transport == "nvme"
  maxSize: 200GB
---
apiVersion: v1alpha1
kind: UserVolumeConfig
name: longhorn
provisioning:
  diskSelector:
    match: disk.transport == "nvme"
  maxSize: 800GB
filesystem:
  type: xfs
```

Longhorn "just works" with volume pointed at `/var/mnt/longhorn`, and upgrades are safe.

The only "meh" I've noticed is that if I apply both patches together, the order of their creation seems to be non-deterministic (which obviously doesn't *really* matter); I currently prefer to bootstrap with the `VolumeConfig` then follow-up by applying the `UserVolumeConfig`.

**Comment by schneid-l** [comment]
Hey,

Happy to see that work is progressing on this great story.
Thanks again for all your hard work!

Is netboot/ramboot support still being considered?

**Comment by smira** [comment]
> Is netboot/ramboot support still being considered?

yes, but not for immediate Talos 1.11 yet, might be some later release.

**Comment by TsengSR** [comment]
@isometry 
> > Anything new on this topic?
> 
> My setup is also built on Turing RK1 with an NVMe per node. YMMV, but I've been very happy with Talos imaged to the eMMC, then EPHEMERAL and Longhorn on the NVMe with the following:
> 
> The only "meh" I've noticed is that if I apply both patches together, the order of their creation seems to be non-deterministic (which obviously doesn't _really_ matter); I currently prefer to bootstrap with the `VolumeConfig` then follow-up by applying the `UserVolumeConfig`.

To be honest, to docs right now are everything but beginner friendly, at least in terms of disk/volume management, which should be one of the most basic and commonly used features. 

Its so hard to get started in when you're new to Talos. Like the above ones (I also seen [this](https://www.talos.dev/v1.10/talos-guides/configuration/disk-management/#system-volume-configuration)), they look like kubernetes manifests, but theres no VolumeConfig/UserVolumeConfig custom resources in kubernetes so obviously they were not meant to be applied with kubectl. 
All talos related configurations int he docs are about machine and cluster configuration via yaml where you put settings under either `machine:`  or `cluster:` but for the volumes there is **no hint** where to put them nor a command for execution.

I litterally spent 7 hours looking how to create a partition for longhorn and fail to even get to an attempt at creating it. Putting it midly its infuriating.

**Comment by isometry** [comment]
> To be honest, to docs right now are everything but beginner friendly, at least in terms of disk/volume management, which should be one of the most basic and commonly used features. 

You're not wrong.

> I litterally spent 7 hours looking how to create a partition for longhorn and fail to even get to an attempt at creating it. Putting it midly its infuriating.

In case you're still looking, I'm unsure if this is the "correct" approach, but what I've done since the introduction of these features is simply append these YAML docs to the end of my main machine config files, to be applied with `talosctl [bootstrap|apply]`.

**Comment by TsengSR** [comment]
@isometry 

> In case you're still looking, I'm unsure if this is the "correct" approach, but what I've done since the introduction of these features is simply append these YAML docs to the end of my main machine config files, to be applied with `talosctl [bootstrap|apply]`.

I tried that and at least it didn't give me an error. But its not changing the partitions. Also not after restart. Also not after resetting the node and applying it. 
Even after reseting and before applying `talosctl get discoveredvolumes -n 192.168.x.x` still shows the 2 TB partition there. 

I can't delete it with `talosctl wipe disk -n 192.168.x.x nvme0n1p6 --drop-partition` while the node is configured because its in use.
> 1 error occurred:
>        * 192.168.x.x: rpc error: code = FailedPrecondition desc = blockdevice "nvme0n1p6" is in use by volume "EPHEMERAL"

I can't delete it with `talosctl wipe disk -n 192.168.x.x nvme0n1p6 --drop-partition` when its resetted. 
>rpc error: code = Unavailable desc = connection error: desc = "error reading server preface: remote error: tls: certificate required"

Also `--insecure` isn't valid on talosctl wipe. 

I'm at loss.

**Comment by isometry** [comment]
> I'm at loss.

@TsengSR : If you're happy to wipe all data, you should be able to use `talosctl reset --wipe-mode all`.

**Comment by Syntax3rror404** [comment]
@TsengSR had also this problem but in my case I have two nvme disks, one dedicated for longhorn and the other for talos linux. In my case I needed to reboot talos and after that wiping the disk with `talosctl wipe disk <nvme> --drop-partition`. Without the reboot the command succeeded but the partition still persists on the nvme. 

Try to disable scheduling on longhorn on this node with eviction enabled, to move your replicas to another node and afterwards remove the disk in the edit node screen in longhorn ui. Then comment out the userconfigmount for the node and apply it and after that reboot your node and wipe the partion down. If the wipe command is done comment userconfigmount in and apply the machineconfig. This helped me in this case.

Edit for reference: https://github.com/siderolabs/talos/issues/11223

**Comment by TsengSR** [comment]
> > I'm at loss.
> 
> [@TsengSR](https://github.com/TsengSR) : If you're happy to wipe all data, you should be able to use `talosctl reset --wipe-mode all`.

Yea, that's fine and worked. Didn't had any stateful services, only etcd, but shouldn't be an issues with 3 controlplanes. 
Altough it also wiped talos of the NVMe, but its quickly flashed back. 

Wish that would options flag would have been documented [here](https://www.talos.dev/v1.10/talos-guides/resetting-a-machine/).

**Comment by Syntax3rror404** [comment]
> > > I'm at loss.
> > 
> > 
> > [@TsengSR](https://github.com/TsengSR) : If you're happy to wipe all data, you should be able to use `talosctl reset --wipe-mode all`.
> 
> Yea, that's fine and worked. Didn't had any stateful services, only etcd, but shouldn't be an issues with 3 controlplanes. Altough it also wiped talos of the NVMe, but its quickly flashed back.
> 
> Wish that would options flag would have been documented [here](https://www.talos.dev/v1.10/talos-guides/resetting-a-machine/).

Yea, I use as workaround always `talosctl` in combination with `--help` to get all available commands, so I dont need to go to the documentaion :D

By the way `--wipe-mode all` is the default. Strange that `talosctl reset --wipe-mode all` changed the behaviour, I wonder if the description of this command outdated?

```
talosctl reset --help
Reset a node

Usage:
  talosctl reset [flags]

Flags:
      --debug                                    debug operation from kernel logs. --wait is set to true when this flag is set
      --graceful                                 if true, attempt to cordon/drain node and leave etcd (if applicable) (default true)
  -h, --help                                     help for reset
      --insecure                                 reset using the insecure (encrypted with no auth) maintenance service
      --reboot                                   if true, reboot the node after resetting instead of shutting down
      --system-labels-to-wipe strings            if set, just wipe selected system disk partitions by label but keep other partitions intact
      --timeout duration                         time to wait for the operation is complete if --debug or --wait is set (default 30m0s)
      --user-disks-to-wipe strings               if set, wipes defined devices in the list
      --wait                                     wait for the operation to complete, tracking its progress. always set to true when --debug is set (default true)
      --wipe-mode all, system-disk, user-disks   disk reset mode (default all)

Global Flags:
      --cluster string       Cluster to connect to if a proxy endpoint is used.
      --context string       Context to be used in command
  -e, --endpoints strings    override default endpoints in Talos configuration
  -n, --nodes strings        target the specified nodes
      --talosconfig string   The path to the Talos configuration file. Defaults to 'TALOSCONFIG' env variable if set, otherwise '$HOME\.talos\config' and '\var\run\secrets\talos.dev\config' in order.
```

But im happy that this worked for you :)

**Comment by sarendipitee** [comment]
> [@TsengSR](https://github.com/TsengSR) had also this problem but in my case I have two nvme disks, one dedicated for longhorn and the other for talos linux. In my case I needed to reboot talos and after that wiping the disk with `talosctl wipe disk <nvme> --drop-partition`. Without the reboot the command succeeded but the partition still persists on the nvme.

Goodness gracious you are a saint. I've been stuck for hours trying to get a non-system drive partitioned and that fixed it within seconds.


<details>
System install was fine. Got talos installed to a small dedicated drive with the `install: ` key. Then I attempted to provision the unformatted nvme drive at `/dev/nvme0n1`

I at first thought I had to have a `disks: ` key in the `machine: ` config so I tried applying that and it immediately rebooted the whole system without warning (thank god it's a test machine). When it came back up I was getting errors about kubelet's volume not being readable and the system failed to initialization k8s. Yikes.

I eventually came to the realization that `disks: ` is deprecated in favor of `UserVolumeConfigs`. So I removed the disks key entirely and re-applied. K8s was able to start back up.

```
apiVersion: v1alpha1
kind: UserVolumeConfig
name: local-path-provisioner
provisioning:
  diskSelector:
    match: disk.transport == 'nvme' 
```

But this was the first mention of applying secondary specs files to a node config that aren't the main `machine:` specs — so I really wasn't sure the intended way of going about this? Do you add it to your machine.yml / controlplane.yml? Is there a special command for applying non-"machine" config? Couldn't find anything about this in the docs and there didn't seem to be any other talosctl related commands so I went with `patch machineconfig --patch @<file>` and that seemed to do the trick (just very confusing that it's not really patching the "machine" config but apply config to a node idk.

This then got me stuck on this error:
```
 user: warning: [2025-06-23T06:32:56.083794732Z]: [talos] volume status {"component":
 "controller-runtime", "controller": "block.VolumeManagerController", "volume": "u-
 local-path-provisioner", "phase": "failed -> failed", "error": "no disks matched for
 volume"}
```

where I went through dozens of attempts of changing the diskSelector using various CELs and bus paths and serial numbers and what have you but no matter what it kept failing to find the disk and I was literally THIS close to giving up on talos forever and going back to CoreOS until I came across this thread and that magically comment to wipe the disk.
 

<summary>My retrospective as a new user of talos if anyone is interested:</summary>

</details>

**Comment by davidgibbons** [comment]
Neither the `reset --wipe-mode all.` or `wipe` command will clear out lvm partitions. Which is a real pain when pulling disks that may have been used for something else previously.  It would be great to have some solution to remove these that didn't come down to pulling the disks out and formatting them in a different system. 

```
❯ talosctl -n <node> wipe disk dm-0 --drop-partition                                                                                                                                                                                                                
❯ talosctl -n <node> wipe disk nvme0n1 --drop-partition                                                                                                                                                                                                             
1 error occurred:
        * <node>: rpc error: code = FailedPrecondition desc = blockdevice "nvme0n1" is in use by disk "dm-0"
❯ talosctl -n <node> wipe disk nvme0n1p3 --drop-partition                                                                                                                                                                                                           
1 error occurred:
        * <node>: rpc error: code = FailedPrecondition desc = blockdevice "nvme0n1p3" is in use by blockdevice "dm-0"
```
I think this is also preventing my machineconfig from using my VolumeConfig to move my EPHEMERAL volume,

**Comment by smira** [comment]
@davidgibbons Talos doesn't manage LVMs at the moment, so you need to "unmount" LVM volume first, as Talos correctly sees and doesn't allow wiping a block device which is in use by LVM.

**Comment by davidgibbons** [comment]
@smira  sure, but just to clarify. "unmount" is in quotes because that's also not something Talos supports. Talos automatically mounted all of these devices and then had zero tools for to remove or wipe them.  

The next node I had a related similar problem where I removed the partition on the device but Talos wouldn't use it because it identified it was ZFS and I had asked for a new volume that was XFS. 

I think in both of these cases if the `wipe` command was made a little more functional it would be a huge increase in the user experience, allow it to run with --insecure,I think would mean i could clear partitions and drives before the system decided to automount them.

**Comment by smira** [comment]
Talos can wipe all volumes, including ZFS, but you'd have to deactivate them first. 

Maintenance mode doesn't mean that volumes will not be activated by the way, Linux way is strange but all LVM volumes are activated automatically on boot, and some CSIs expect that.

**Comment by davidgibbons** [comment]
@smira am I missing some option or command to deactivate storage? Because again the only functional solution that I found was to rip the drives out and put them in an system that had a `parted` command.

**Comment by smira** [comment]
> [@smira](https://github.com/smira) am I missing some option or command to deactivate storage? Because again the only functional solution that I found was to rip the drives out and put them in an system that had a `parted` command.

If you have questions, it's better to open a discussion and ask for help, this issue is not related to wiping LVMs.

**Comment by theAkito** [comment]
I skimmed through the comments and am still asking myself:
Is there a valid workaround or not?
Valid workaround would be to not be forced to reset Talos entirely, just for that `EPHEMERAL` to become reasonable. Is an approach with that goal in mind available?
Or is, as of now, only a reset of entire Talos a workaround?

**Comment by smira** [comment]
Closing this issue, as it became out of context, the next piece of work is tracked in https://github.com/siderolabs/talos/issues/10928
